src/charm.py

```diff
@@ -1,103 +1,168 @@
-        environment configuration for your specific workload.
+        Installs apt packages required to build the image.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+        self.unit.status = ops.MaintenanceStatus("Setting up Builder.")
+        proxy.setup(proxy=state.ProxyConfig.from_env())
+        self._create_success_callback_script()
+        self._create_failed_callback_script()
+        init_config = state.BuilderInitConfig.from_charm(self)
+        # The current CLI version only accepts relative pathing
+        init_config.run_config.callback_script = (
+            init_config.run_config.callback_script.relative_to(builder.UBUNTU_HOME)
+        )
+        builder.initialize(init_config=init_config)
+        self.unit.status = ops.ActiveStatus("Waiting for first image.")
+        builder.run(config=init_config.run_config)
+
+    def _create_success_callback_script(self) -> None:
+        """Create callback script to propagate images."""
+        charm_dir = os.getenv("JUJU_CHARM_DIR")
+        cur_env = {
+            "JUJU_DISPATCH_PATH": f"hooks/{BUILD_SUCCESS_EVENT_NAME}",
+            "JUJU_MODEL_NAME": self.model.name,
+            "JUJU_UNIT_NAME": self.unit.name,
+            OPENSTACK_IMAGE_ID_ENV: "$OPENSTACK_IMAGE_ID",
+        }
+        env = " ".join(f'{key}="{val}"' for (key, val) in cur_env.items())
+        script_contents = f"""#! /bin/bash
+OPENSTACK_IMAGE_ID="$1"
+
+/usr/bin/juju-exec {self.unit.name} {env} {charm_dir}/dispatch
+"""
+        state.SUCCESS_CALLBACK_SCRIPT_PATH.write_text(script_contents, encoding="utf-8")
+        state.SUCCESS_CALLBACK_SCRIPT_PATH.chmod(0o755)
+
+    def _create_failed_callback_script(self) -> None:
+        """Create callback script to propagate images."""
+        charm_dir = os.getenv("JUJU_CHARM_DIR")
+        cur_env = {
+            "JUJU_DISPATCH_PATH": f"hooks/{BUILD_SUCCESS_EVENT_NAME}",
+            "JUJU_MODEL_NAME": self.model.name,
+            "JUJU_UNIT_NAME": self.unit.name,
+        }
+        env = " ".join(f'{key}="{val}"' for (key, val) in cur_env.items())
+        script_contents = f"""#! /bin/bash
+OPENSTACK_IMAGE_ID="$1"
+
+/usr/bin/juju-exec {self.unit.name} {env} {charm_dir}/dispatch
+"""
+        state.FAILED_CALLBACK_SCRIPT_PATH.write_text(script_contents, encoding="utf-8")
+        state.FAILED_CALLBACK_SCRIPT_PATH.chmod(0o755)
+
+    @charm_utils.block_if_invalid_config(defer=False)
```

Comment:
Out of curiosity: Couldn't this be done similarly to this one?
https://github.com/canonical/synapse-operator/blob/main/src/charm_state.py#L64
So we could start having a pattern on how to block a charm in case of an invalid configuration.

Reply:
Oh this is pretty cool, I've taken mine from Paas-app-charmer, https://github.com/canonical/paas-app-charmer/blob/main/paas_app_charmer/_gunicorn/charm_utils.py#L35.
I do like the way the way synapse-operator does to inject the state :D
But one thing though, is that I no longer use a big "charm state" but smaller dataclass objects (e.g. BuilderRunConfig) and load them at the time they are needed. Hence if I do inject those, I will need to inject multiple states which may not be desirable (I wouldn't want to couple the charm_util decorators to the state object).
I think the pattern is super helpful for charms with more generic single charm state instance, thank you for the suggestion!